jvector-base/src/main/java/io/github/jbellis/jvector/pq/ProductQuantization.java

```diff
@@ -307,64 +328,66 @@ public static ProductQuantization load(RandomAccessReader in) throws IOException
             subvectorSizes[i] = new int[2];
             int size = in.readInt();
             subvectorSizes[i][0] = size;
-            offset += size;
             subvectorSizes[i][1] = offset;
+            offset += size;
         }
 
         int clusters = in.readInt();
-        float[][][] codebooks = new float[M][][];
+        VectorFloat<?>[] codebooks = new VectorFloat<?>[M];
         for (int m = 0; m < M; m++) {
-            float[][] codebook = new float[clusters][];
-            for (int i = 0; i < clusters; i++) {
-                int n = subvectorSizes[m][0];
-                float[] centroid = new float[n];
-                in.readFully(centroid);
-                codebook[i] = centroid;
-            }
+            VectorFloat<?> codebook = vectorTypeSupport.readFloatVector(in, clusters * subvectorSizes[m][0]);
             codebooks[m] = codebook;
         }
 
-        return new ProductQuantization(codebooks, globalCentroid);
+        return new ProductQuantization(clusters, subvectorSizes, codebooks, globalCentroid);
     }
 
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         ProductQuantization that = (ProductQuantization) o;
-        return M == that.M
+        if (M == that.M
                && originalDimension == that.originalDimension
-               && Arrays.deepEquals(codebooks, that.codebooks)
-               && Arrays.equals(globalCentroid, that.globalCentroid)
-               && Arrays.deepEquals(subvectorSizesAndOffsets, that.subvectorSizesAndOffsets);
+               && Objects.equals(globalCentroid, that.globalCentroid)
+               && Arrays.deepEquals(subvectorSizesAndOffsets, that.subvectorSizesAndOffsets)) {
+            for (int m = 0; m < M; m++) {
```

Comment:
Huh, I don't remember why I didn't include codebooks before.

But I think you can use Arrays.deepEquals on codebooks?